modules/core/04-channel/keeper/upgrade_test.go

```diff
@@ -2150,3 +2150,79 @@ func (suite *KeeperTestSuite) TestSyncUpgradeSequence() {
 		})
 	}
 }
+
+func (suite *KeeperTestSuite) TestChanUpgradeCrossingHelloWithHistoricalProofs() {
+	var path *ibctesting.Path
+
+	testCases := []struct {
+		name     string
+		malleate func()
+		expError error
+	}{
+		{
+			"success",
+			func() {},
+			nil,
+		},
+		{
+			"counterparty (chain B) has already progressed to ACK step",
+			func() {
+				err := path.EndpointB.ChanUpgradeAck()
+				suite.Require().NoError(err)
+			},
+			types.ErrInvalidChannelState,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+
+			path = ibctesting.NewPath(suite.chainA, suite.chainB)
+			suite.coordinator.Setup(path)
+
+			path.EndpointA.ChannelConfig.ProposedUpgrade.Fields.Version = mock.UpgradeVersion
+			path.EndpointB.ChannelConfig.ProposedUpgrade.Fields.Version = mock.UpgradeVersion
+
+			err := path.EndpointA.ChanUpgradeInit()
+			suite.Require().NoError(err)
+
+			err = path.EndpointB.ChanUpgradeInit()
+			suite.Require().NoError(err)
+
+			suite.coordinator.CommitBlock(suite.chainA, suite.chainB)
+
+			err = path.EndpointB.UpdateClient()
+			suite.Require().NoError(err)
+
+			// TODO: QueryChannelUpgradeProof queries the counterparty (EndpointA), this feels unintuitive
+			historicalChannelProof, historicalUpgradeProof, proofHeight := path.EndpointB.QueryChannelUpgradeProof()
```

Comment:
Right now, `path.EndpointA.QueryChannelUpgradeProof()` is querying the counterparty of endpointB (i.e. chain A). 

I feel like this is slightly unintuitive as it reads as if you are querying endpointB for the proof but in reality whats happening is we are querying for proofs which we would like to verify on endpointB(chainB).

I'd like to adapt this in a follow up PR.

Reply:
Opened #5147

Reply:
addressed [1c975b9](https://github.com/cosmos/ibc-go/pull/5143/commits/1c975b9ba0dfdb4293a2c02f9e1a0822bf4d063c)